CHORE: Setup IaC via Pulumi

furiousdom opened the pull request:
## Overview

- Added IaC via Pulumi leveraging [@studion/infra-code-blocks](https://github.com/ExtensionEngine/infra-code-blocks) package
  - Tested that everything works by:
    - Checking AWS console instances are up and running
    - Connecting to the database via SSM
- The app can't be reached via browser (as far as dev knows), because there isn't a DNS route set up for it
  - This is intended because the app currently isn't in a good enough state for this to matter
  - Also, the main point of this work is to set up CI/CD pipeline to make deployment easier later, and DNS routing isn't needed for that, only that infrastructure is deployed via Pulumi

## Reviewer Tips

- Huge line diff is because of `pacakge-lock.json`

## Setup and Deployment

### Setup

Note: All commands are run in terminal
**IMPORTANT NOTE**: Working directory for commands needs to be `<project-root>/infrastructure`

1. Run `pulumi stack init <stack-name>
2. Run `pulumi config set aws:region <aws-region>`
3. Run `pulumi config set --secret <pulumi-project-name>:dbUsername <db-name>`

### Deployment

- In terminal, run `pulumi up`, then at confirmation prompt select `yes`

furiousdom committed:
Initialize Pulumi project

- The pulumi project won't create anything

furiousdom committed:
Delete `Pulumi.dev.yaml`

- This is a prep step, as the file will be set to be ignored by git in future commits

furiousdom committed:
Make `git` ignore Pulumi dev config file

furiousdom committed:
Remove bucket from Pulumi project

furiousdom committed:
Add demo VPC project to Pulumi

furiousdom committed:
Add starting EC2 & RDS instances via Pulumi

furiousdom committed:
Add `@studion/infra-code-blocks` dependency

furiousdom committed:
Change Pulumi project to leverage `@studion/infra-code-blocks` package

- Change main components of the infrastructure from EC2 and RDS instances to ECS Fargate, ECR, and RDS
- Add template to Pulumi config for non prod config
- "dbUsername" needs to be added via "pulumi config set --secret <pulumi-project-name>:dbUsername <some-db-name>

abasic reviewed (commented):
I won't merge this yet because I want you to look into if you can access the application through an IP. I did a quick look through `infra-code-blocks` and it might have set up a load balancer so it should have an IP you can directly ping.

furiousdom committed:
Fix missing `PORT` environment variable

furiousdom committed:
Add missing `NODE_ENV` environment variable

furiousdom committed:
Add `POSTGRES_SSL_ENABLED` variable to server config

furiousdom committed:
Rename `POSTGRES_SSL_ENABLED` to more generic `IS_DATABASE_SSL_ENABLED`

furiousdom committed:
Fix `DATABASE_PORT` variable being NaN value

- Also fix issue with "DATABASE_PASSWORD" variable being something it shouldn't be
- Both issues stem from the same root problem, not resolving Pulumi variable values

furiousdom committed:
Use `IS_DATABASE_SSL_ENABLED` env in a meaningful way

furiousdom committed:
Add `healthcheck` endpoint to application and add it to pulumi

- This fixes an issue in AWS infrastructure where the app is constantly being restarted

furiousdom committed:
Fix problem with app unavailability through browser

- When hitting either Elastic IP, or load balancer DNS address, a timeout error occurs

furiousdom committed:
Fix Internal Server Error

furiousdom commented:
You were right, the app didn't work properly, and there were a couple of other things wrong as well. Now infrastructure works fine and the app can be accessed through load balancer DNS address, but there is still one issue. The app can't be accessed through Elastic IP.
